lib/charms/mongodb/v0/upgrade_helpers.py

```diff
@@ -0,0 +1,464 @@
+#!/usr/bin/env python3
+"""Substrate agnostic manager for handling MongoDB in-place upgrades."""
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+import abc
+import logging
+import secrets
+import string
+from typing import List, Tuple
+
+from charms.mongodb.v0.mongo import MongoConfiguration
+from charms.mongodb.v1.mongodb import MongoDBConnection
+from charms.mongodb.v1.mongos import MongosConnection
+from ops.charm import CharmBase
+from ops.framework import Object
+from pymongo.errors import OperationFailure, PyMongoError, ServerSelectionTimeoutError
+from tenacity import Retrying, retry, stop_after_attempt, wait_fixed
+
+from config import Config
+
+# The unique Charmhub library identifier, never change it
+LIBID = "7f59c0d1c6f44d1993c57afa783e205c"
+
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+SHARD_NAME_INDEX = "_id"
+WRITE_KEY = "write_value"
+ROLLBACK_INSTRUCTIONS = "To rollback, `juju refresh` to the previous revision"
+
+
+# BEGIN: Useful classes
+class AbstractUpgrade(abc.ABC):
```

Comment:
This class here is mainly for typing.
But it could also be used for the same use as `GenericMongoDBUpgrade` to embed the methods that we want to share between the VM and k8s charm.
For now it's an empty class for typing, later it can be extended when we have new upgrades and if there are some methods and patterns that can be shared.